lib/stacktraces.js

```diff
@@ -0,0 +1,416 @@
+'use strict'
+
+// Handling for translating `<Error instance>.stack` to the
+// `{error.exception,span}.stacktrace` object passed to APM server.
+// https://github.com/elastic/apm-server/blob/master/docs/spec/v2/error.json#L633
+//
+// A lot of this derived from https://github.com/watson/stackman but was
+// moved internal to allow breaking compat for perf work.
+
+var fs = require('fs')
+var path = require('path')
+
+const asyncCache = require('async-cache')
+const afterAllResults = require('after-all-results')
+const errorCallsites = require('error-callsites')
+const errorStackParser = require('error-stack-parser')
+const loadSourceMap = require('load-source-map')
+const LRU = require('lru-cache')
+
+const fileCache = asyncCache({
+  max: 500, // fileCacheMax
+  load: function (file, cb) {
+    fs.readFile(file, { encoding: 'utf8' }, function (err, data) {
+      if (err) {
+        cb(err)
+        return
+      }
+      cb(null, data.split(/\r?\n/))
+    })
+  }
+})
+
+// Note: This is not calling `sourcemap.destroy()`
+// (https://github.com/mozilla/source-map#sourcemapconsumerprototypedestroy) and
+// it probably should. async-cache does not provide a mechanism to do something
+// with evicted values.
+const sourceMapCache = asyncCache({
+  max: 500, // sourceMapCacheMax
+  load: function (filename, cb) {
+    if (!filename) {
+      cb(null, null)
+      return
+    }
+    loadSourceMap(filename, function onSourceMap (err, sourcemap) {
+      // Translate sourcemap===undefined to null, because 'async-cache'
+      // treats `undefined` as a cache miss. Without this there is no
+      // caching for files that have no sourcemap (the common case).
+      cb(err, sourcemap || null)
+    })
+  }
+})
+
+const frameCache = new LRU({ max: 1000 })
```

Comment:
> Do we have any back of the napkin numbers on how much this will could, potentially, increase overall memory use for the agent?

Good Q. Given an example frame is ~500 bytes when JSON serialized... I'd have hoped it would be as simple as saying "it'll take up ~500kB of added memory". However playing around a bit with the implementation and applying the load (500 req/s) to pathosapp.js, it isn't so simple. The numbers below are "RES" (resident size) from watching `htop` during a load run. Note that RES oscillated plus/minus 5-10MB during runs.

- using the current agent release (v3.15.0): RES=~122M
- using the current PR adjusted to comment out `frameCache` usage: RES=~144M
- using the current PR adjusted to bust the cacheKey so we hit 1000 entries *and* always have a cache miss: RES=~155M
- using the current PR adjusted to have a `frameCache = new LRU({ max: 1 })`: RES=~155M
- using the current PR adjusted with the patch just below to hit 1000 entries, but still cache effectively: RES=~123M
- ditto, but using node 16: RES=~110MB (10-12MB less than node v14!)

```diff
diff --git a/lib/stacktraces.js b/lib/stacktraces.js
index 2809080f..aa7cad9b 100644
--- a/lib/stacktraces.js
+++ b/lib/stacktraces.js
@@ -241,7 +241,8 @@ function frameFromCallSite (log, callsite, cwd, sourceLinesAppFrames, sourceLine
   const colno = callsite.getColumnNumber()

   // Caching
-  const cacheKey = [filename, lineno, colno, sourceLinesAppFrames, sourceLinesLibraryFrames].join(':')
+  let cacheKey = [filename, lineno, colno, sourceLinesAppFrames, sourceLinesLibraryFrames].join(':')
+  if (frameCache.itemCount < 1000) cacheKey += Math.random().toString()
   const cachedFrame = frameCache.get(cacheKey)
   if (cachedFrame !== undefined) {
     frameCacheStats.hits++
@@ -251,6 +252,7 @@ function frameFromCallSite (log, callsite, cwd, sourceLinesAppFrames, sourceLine
   function cacheAndCb (frame) {
     frameCacheStats.misses++
     frameCache.set(cacheKey, frame)
+    frameCacheStats.itemCount = frameCache.itemCount
     cb(null, frame)
   }

```

My conclusions:
- It's complicated.
- A heap profile might help given more concrete data, but I'm not proposing to do that right now.
- I *believe* the memory usage of that cache is mostly lost in the noise of the changes in behaviour (handling concurrent operations differently under load) with the effective caching. Recall that when getting cache hits the CPU usage dropped 7-10%

Reply:
> Also, more generally, any thoughts on whether the `lru-cache` module is what we want to use here long term, or if this seems like something to target for simplification in a future PR?

For what it's worth, I *did* do a load run with a simple `const frameCache = new Map()` (which didn't do any `max: 1000` handling) vs. using `new LRU({ max: 1000 })` and for CPU usage or flamegraph I saw no change. So CPU performance-wise, I think lru-cache is fine.

Also, for what it's worth, I've been using lru-cache for almost a decade and I'm pretty happy with it. For a while I had https://github.com/trentm/node-expiring-lru-cache which added expiry support to `lru-cache`. Then isaacs added `maxAge` support to `lru-cache` making my module unnecessary.  Happy to re-evaluate later at some point, but I think it'll be low prio. `lru-cache` is pretty battle tested.

Reply:
👍